### DISCUSSION TITLE
Run strongswan as a regular user

### DISCUSSION THREAD

**Discussion by rootBiao** [strongswan/strongswan#2178] [General]
![image](https://github.com/strongswan/strongswan/assets/87700793/31ec9636-842b-4fb6-910b-0e4534a87f25)
Do I have to run strongswan as root user？

**Comment by rootBiao** [comment]
https://docs.strongswan.org/docs/5.9/install/reducedPrivileges.html#:~:text=To%20use%20a%20non%2Droot%20user%2C%20you%E2%80%99ll%20need%20to%20create%20a%20new%20user%20and%20group%20for%20strongSwan%2C%20e.g.

I followed this approach and ran the program as a non root user, but the following error occurred.
![image](https://github.com/strongswan/strongswan/assets/87700793/8f13772c-6fb9-460b-b926-803968d179f3)

**Reply by tobiasbrunner** [reply]
The daemon needs the NET_ADMIN capability to communicate via XFRM sockets with the Linux kernel. When switching users, it needs a way to keep that capability, so you have to use `--with-capabilities` (see the page you linked for details).

**Reply by rootBiao** [reply]
![image](https://github.com/strongswan/strongswan/assets/87700793/58347695-d21c-4f26-ac19-da3c9f918a92)
like this?

**Reply by tobiasbrunner** [reply]
No. Please read the documentation.

**Reply by rootBiao** [reply]
But when I execute the ipsec start command in the Non-Root user's state, I get the following error.
![image](https://github.com/strongswan/strongswan/assets/87700793/b9a33015-c5ca-47b9-a383-48c375f7a837)
![image](https://github.com/strongswan/strongswan/assets/87700793/0bec5352-92d3-44fc-b206-3093dfbb93ea)

**Reply by tobiasbrunner** [reply]
First, `ipsec` is a legacy script, it has nothing to do with the IKE daemon directly. And the daemon has to start as root (unless you do some stuff with assigning capabilities to the executable), it then switches the user/group and keeps only the capabilities it needs.

**Reply by rootBiao** [reply]
![image](https://github.com/strongswan/strongswan/assets/87700793/bbe969d9-0625-42dc-adb0-d061e045872b)
Can I include the definition of macro STARTER_ALLOW_NON_ROOT during compilation?

**Reply by tobiasbrunner** [reply]
Sure, if you want to use starter, but there isn't really any reason to do so.

**Reply by rootBiao** [reply]
![image](https://github.com/strongswan/strongswan/assets/87700793/ed94968b-3839-4894-a099-0b1382f299fa)
![image](https://github.com/strongswan/strongswan/assets/87700793/1101a9f9-8458-4fe9-b84c-d4001278a3f1)


![image](https://github.com/strongswan/strongswan/assets/87700793/2d0de4bb-dbcd-4fc4-9e11-738826ae9f5f)


what permissions do I need to grant executable files when I want to start as a non root user？
I just want to obtain the relevant key information for encryption and decryption of child SA through strongswan.

**Reply by tobiasbrunner** [reply]
The daemon needs at least the NET_ADMIN capability. As I said before, it is usually started as root and then switches to the configured user/group and drops all unneeded capabilities. And if the container is already restricted to a set of capabilities I don't know if running it as a non-root user actually gains you much (in particular if you configured the daemon to switch to a non-root user). Anyway, starting as non-root user is theoretically possible but then the capabilities have to be obtained some other way (e.g. as file-level capabilities via `setcap`).

**Reply by rootBiao** [reply]
On some platforms, we are assigned a regular user, not a root user.
